From 7d689a1efe97827f8e2b188b3dcee8688e1fe33f Mon Sep 17 00:00:00 2001
From: Felix Kuehling <Felix.Kuehling@amd.com>
Date: Thu, 16 Nov 2017 12:41:07 -0500
Subject: [PATCH 3282/5855] drm/amdgpu: Add back invalidate_page MMU notifier

The invalidate_page notifier is still needed for DKMS installations
on older kernels without Jerome Glisse's patch series that removed
it. Jerome's patch series landed upstream in 4.14.

In 4.11 the invalidate_page notifier was broken because it was
called under a spinlock. Between 4.11 and 4.14 we prefer not to
implement it to avoid deadlocks, at the cost of potential subtle
consistency issues.

Change-Id: Ie7f94a39de819e83b3aad69bf477fb06019ea982
Signed-off-by: Felix Kuehling <Felix.Kuehling@amd.com>
Signed-off-by: Raveendra Talabattula <raveendra.talabattula@amd.com>
---
 drivers/gpu/drm/amd/amdgpu/amdgpu_mn.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_mn.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_mn.c
index bc8187f..1093dfc 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_mn.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_mn.c
@@ -204,7 +204,7 @@ static void amdgpu_mn_invalidate_page(struct mmu_notifier *mn,
 	struct amdgpu_mn *rmn = container_of(mn, struct amdgpu_mn, mn);
 	struct interval_tree_node *it;
 
-	down_write(&rmn->lock);
+	amdgpu_mn_read_lock(rmn);
 
 	it = interval_tree_iter_first(&rmn->objects, address, address);
 	if (it) {
@@ -214,7 +214,7 @@ static void amdgpu_mn_invalidate_page(struct mmu_notifier *mn,
 		amdgpu_mn_invalidate_node(node, address, address);
 	}
 
-	up_write(&rmn->lock);
+	amdgpu_mn_read_unlock(rmn);
 }
 
 /**
@@ -322,7 +322,7 @@ static void amdgpu_mn_invalidate_page_hsa(struct mmu_notifier *mn,
 	struct amdgpu_mn *rmn = container_of(mn, struct amdgpu_mn, mn);
 	struct interval_tree_node *it;
 
-	down_write(&rmn->lock);
+	amdgpu_mn_read_lock(rmn);
 
 	it = interval_tree_iter_first(&rmn->objects, address, address);
 	if (it) {
@@ -340,7 +340,7 @@ static void amdgpu_mn_invalidate_page_hsa(struct mmu_notifier *mn,
 		}
 	}
 
-	up_write(&rmn->lock);
+	amdgpu_mn_read_unlock(rmn);
 }
 
 static const struct mmu_notifier_ops amdgpu_mn_ops[] = {
-- 
2.7.4

